patches/server/1008-Optimise-general-POI-access.patch

```diff
@@ -894,58 +899,67 @@ index 6f2c7baea0d1ac7813c7b85e1f5558573745762c..12a7aaeaa8b4b788b620b1985591c3b9
+-            .min(Comparator.comparingDouble(poi -> poi.getPos().distSqr(pos)))
+-            .map(poi -> Pair.of(poi.getPoiType(), poi.getPos()));
 +        // Paper start - re-route to faster logic
 +        return Optional.ofNullable(io.papermc.paper.util.PoiAccess.findClosestPoiDataTypeAndPosition(
 +            this, typePredicate, null, pos, radius, radius * radius, occupationStatus, false
 +        ));
 +        // Paper end - re-route to faster logic
      }
  
-     public Optional<BlockPos> findClosest(Predicate<Holder<PoiType>> typePredicate, Predicate<BlockPos> posPredicate, BlockPos pos, int radius, PoiManager.Occupancy occupationStatus) {
--        return this.getInRange(typePredicate, pos, radius, occupationStatus).map(PoiRecord::getPos).filter(posPredicate).min(Comparator.comparingDouble((blockPos2) -> {
--            return blockPos2.distSqr(pos);
--        }));
+     public Optional<BlockPos> findClosest(
+         Predicate<Holder<PoiType>> typePredicate, Predicate<BlockPos> posPredicate, BlockPos pos, int radius, PoiManager.Occupancy occupationStatus
+     ) {
+-        return this.getInRange(typePredicate, pos, radius, occupationStatus)
+-            .map(PoiRecord::getPos)
+-            .filter(posPredicate)
+-            .min(Comparator.comparingDouble(blockPos2 -> blockPos2.distSqr(pos)));
 +        // Paper start - re-route to faster logic
 +        BlockPos ret = io.papermc.paper.util.PoiAccess.findClosestPoiDataPosition(this, typePredicate, posPredicate, pos, radius, radius * radius, occupationStatus, false);
 +        return Optional.ofNullable(ret);
 +        // Paper end - re-route to faster logic
      }
  
      public Optional<BlockPos> take(Predicate<Holder<PoiType>> typePredicate, BiPredicate<Holder<PoiType>, BlockPos> biPredicate, BlockPos pos, int radius) {
--        return this.getInRange(typePredicate, pos, radius, PoiManager.Occupancy.HAS_SPACE).filter((poi) -> {
--            return biPredicate.test(poi.getPoiType(), poi.getPos());
--        }).findFirst().map((poi) -> {
+-        return this.getInRange(typePredicate, pos, radius, PoiManager.Occupancy.HAS_SPACE)
+-            .filter(poi -> biPredicate.test(poi.getPoiType(), poi.getPos()))
+-            .findFirst()
+-            .map(poi -> {
+-                poi.acquireTicket();
+-                return poi.getPos();
+-            });
 +        // Paper start - re-route to faster logic
 +        final @javax.annotation.Nullable PoiRecord closest = io.papermc.paper.util.PoiAccess.findClosestPoiDataRecord(
 +            this, typePredicate, biPredicate, pos, radius, radius * radius, Occupancy.HAS_SPACE, false
 +        );
 +        return Optional.ofNullable(closest).map(poi -> {
 +            // Paper end - re-route to faster logic
-             poi.acquireTicket();
-             return poi.getPos();
-         });
++            poi.acquireTicket();
++            return poi.getPos();
++        });
```

Comment:
move .map down a line and put the Paper end commend above that (and revert indent change on these 3 lines)